monocle/gitlab/mergerequest.py

```diff
@@ -0,0 +1,545 @@
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+from dataclasses import dataclass
+from time import sleep
+from typing import Callable, List, Optional, Union
+
+from monocle.basecrawler import BaseCrawler, RawChange
+from monocle.db.db import (
+    Change,
+    Commit,
+    Event,
+    File,
+    SimpleFile,
+    change_or_event_to_dict,
+)
+from monocle.gitlab.graphql import GitlabGraphQLQuery, RequestTimeout
+from monocle.ident import IdentsConfig, create_ident
+from monocle.utils import is8601_to_dt
+
+name = "gitlab_crawler"
+help = "GitLab Crawler to fetch MRs events"
+
+MAX_TRY = 3
+MAX_BULK_SIZE = 100
+REDUCE = 2
+AUGMENT = 1.1
+
+
+class TokenGetter:
+    def __init__(
+        self,
+        url,
+        group,
+        token: Optional[str] = None,
+    ) -> None:
+        self.url = url
+        self.group = group
+        self.token = token
+
+    def get_token(self):
+        if self.token:
+            return self.token, {"contents": "read"}
+        else:
+            raise RuntimeError("TokenGetter need a Token")
+
+    def can_read_commit(self) -> bool:
+        return "contents" in self.get_token()[1].keys()
+
+
+@dataclass
+class GitLabCrawlerArgs(object):
+    updated_since: str
+    loop_delay: int
+    command: str
+    url: str
+    group: str
+    repository: str
+    db: object
+    idents_config: IdentsConfig
+
+
+class MRsFetcher(BaseCrawler):
+
+    log = logging.getLogger(__name__)
+
+    def __init__(
+        self,
+        gql: GitlabGraphQLQuery,
+        url: str,
+        group: str,
+        idents_config: IdentsConfig,
+        repository: str,
+    ):
+        self.gql = gql
+        self.size = MAX_BULK_SIZE
+        self.url = url
+        self.group = group
+        self.repository = repository
+        self.idents_config = idents_config
+
+    def get_mr_query(self, include_commits=True):
+        commits = """
+          edges {
+            node {
+              commit {
+                oid
+                pushedDate
+                authoredDate
+                committedDate
+                additions
+                deletions
+                message
+                author {
+                  user {
+                    login
+                  }
+                }
+                committer {
+                  user {
+                    login
+                  }
+                }
+              }
+            }
+          }
+      """
```

Comment:
is this ok with not having a `noqa` section?

Reply:
It is, as long as none of the linters and formatters complain